generateBins.py

```diff
@@ -25,69 +13,46 @@
 
 
 # TODO Still need a good way of supplying these
-PFPR_FILE       = "rwa_pfpr2to10.asc"
+PFPR_FILE = "rwa_pfpr2to10.asc"
 POPULATION_FILE = "rwa_population.asc"
 
-
 # TODO Determine the bins computationally
 # Following bins are for Rwanda
 POPULATION_BINS = [2125, 5640, 8989, 12108, 15577, 20289, 27629, 49378, 95262, 286928]
 
-# getting data ready for binning
-# data should be 1-dimensional array, python list or iterable
-myArray  = np.loadtxt("GIS\\rwa_population.asc", skiprows = 6)
-print (type(myArray))
-print(myArray.ndim)
-print(myArray.shape)
-print(myArray.size)
-array_1d = myArray.flatten()
-print (array_1d)
-print(array_1d.ndim)
-print(array_1d.shape)
-print(array_1d.size)
-#gvf = 0.0
-#nclasses = 20
-
-def goodness_of_variance_fit(array, classes):
-    # get the break points
-    classes = jenkspy.jenks_breaks(array, classes)
-
-    # do the actual classification
-    classified = np.array([classify(i, classes) for i in array])
-
-    # max value of zones
-    maxz = max(classified)
-
-    # nested list of zone indices
-    zone_indices = [[idx for idx, val in enumerate(classified) if zone + 1 == val] for zone in range(maxz)]
-
-    # sum of squared deviations from array mean
-    sdam = np.sum((array - array.mean()) ** 2)
-
-    # sorted polygon stats
-    array_sort = [np.array([array[index] for index in zone]) for zone in zone_indices]
-
-    # sum of squared deviations of class means
-    sdcm = sum([np.sum((classified - classified.mean()) ** 2) for classified in array_sort])
-
-    # goodness of variance fit
-    gvf = (sdam - sdcm) / sdam
-
-    return gvf
 
-
-def process(configuration, gisPath = ""):
+def process(configuration, gisPath=""):
     # Load the configuration
     cfg = load_configuration(configuration)
+    filename = os.path.join(gisPath, PFPR_FILE)
 
     # TODO Add the stuff for the population bins!
+    # GVF Implementation
+    binning = bin_asc(filename)
+    #gvf = goodness_of_variance_fit(binning, 14)
+    #print("The value of Goodness of Variance fit is:", gvf)
+    #if gvf < 0.7:
+    #    print("Warning: GVF too low")
+
+    list_one = []
+    list_two = []
+    for x in range(5, 26):
+        GVF = goodness_of_variance_fit(binning, x)
+        if GVF < 0.7:
+            print("Warning: GVF too low")
+
+        if GVF >= 0.94:
```

Comment:
Ditto above comment, but also why is 0.94 special. Remember, that this script is going to be used for more than just Rwanda. There's also the Burkina Faso data we currently have, and in the future we will be looking at more countries.

Reply:
Would be it okay if we use `0.9`? Will it be good for all the data we have?

Reply:
@PakhiAgarwal Using a single value for all data is always very tricky, so some sort of heuristic is usually a much better route since they tend to be much more flexible. Recall the comment earlier about stopping the fitting when `abs(gvf(previous) - gvf(current)) <= 0.1`? That's likely a good spot to stop as a basic heuristic and also has the advantage when we a good fit can be found quickly we can go ahead and stop running. For a larger countries with more cells that can be a very valuable feature as well.

Reply:
I thought about using this, but in our current dataset the number of bins will be 7 or 8, this value will be even more less for other datasets maybe 4 or 5.
Would it be good for our work?

Reply:
@PakhiAgarwal Correct, for other data sets the value might be lower. You might want to use a range from 2 to 30 or so just to keep things simple.